### DISCUSSION TITLE
Another CORS issue

### DISCUSSION THREAD

**Discussion by GoldManeitor** [vercel/community#5714] [Help]
### Summary

Hi! yes, i'm the guy who couldn't  deploy a serverless function... well, a very kind user help me and I could deployed succesfully. 

But, now, I have an another problem...CORS :'v

I'll tried every step of Vercel documentation ( [adding CORS headers in the Serverless Function](https://vercel.com/guides/how-to-enable-cors#enabling-cors-in-a-single-node.js-serverless-function) ) and the problem persists. 

I'll show you my code, it only sends emails:

 ```
 import { VercelRequest, VercelResponse } from '@vercel/node';
  import nodemailer from 'nodemailer';
  
  export default async (req: VercelRequest, res: VercelResponse) => {
    try {
      res.setHeader('Access-Control-Allow-Credentials', 'true');
      res.setHeader('Access-Control-Allow-Origin', '*');
      res.setHeader('Access-Control-Allow-Methods', 'GET,OPTIONS,PATCH,DELETE,POST,PUT');
      res.setHeader(
        'Access-Control-Allow-Headers',
        'X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version'
      );
  
      const { email, subject, message } = req.body;
  
      if (!email || !subject || !message) {
        return res.status(400).json({ error: 'Necessary data is missing.' });
      }
  
      // Configuring Nodemailer transport
      const transporter = nodemailer.createTransport({
        host: 'resend.host',
        secure: true,
        port: resend.port,
        auth: {
          user: 'resend.user',
          pass: `${process.env.MY_RESEND_API_KEY}`,
        },
      });
  
      // Email Options
      const mailOptions = {
        from: 'myemail@mydomain.com',
        to: email,
        subject: subject,
        html: message,
      };
  
      // Send the email
      const info = await transporter.sendMail(mailOptions);
  
      console.log('Email sent successfully:', info.messageId);
  
      res.status(200).json({ success: true });
    } catch (error) {
      console.error('Failed to send email:', error);
      res.status(500).json({ error: 'Internal Server Error.' });
    }
  };

```
and this is the vercel.json:

```
{
  "functions": {
    "api/*.*": {
      "memory": 1024
    }
  }
}
```

I think the code its ok, even I tried to add CORS into the function but didn't work too. This is the message that console throws:

> Cross-Origin Request Blocked: Same Origin Policy does not allow reading remote resources at https://www.myfunctionlink.vercel.app/folder/file. (Reason: Unsuccessful CORS request). Status code: (null).

> Error de red: TypeError: NetworkError when attempting to fetch resource.

The same message appears every time, I don't know how to solve it, backend is not my strong point :c
Thanks for your time!

### Steps to Reproduce

- create a project

- install dependecies (vercel, node, typescript, nodemailer, resend etc)

- create a folder

- create a .ts file

- create a similar function like the above

- add the above vercel.json code to vercel.json file, after created it

- create a vercelignore file for enviroment variables

- create a project in Vercel page

- add the project, configure enviroment variables and build

- run it fetching from frontend

**Comment by levilansing** [comment]
Given your function, I'm assuming you're sending "Content-Type": "application/json", which your browser includes in the OPTIONS request. An OPTIONS request has an empty body, but for some reason Vercel helpers are trying to parse the non-existent body, which throws an internal error that gets swallowed and then it returns a 500 internal server error.

At least this is my problem. If I remove the content-type from the request, it works fine, except now I have to manually read the body stream and parse the JSON myself.

This is a serious problem when trying to build a CORS compatible API. I have not found anyway around it besides never sending content type headers
